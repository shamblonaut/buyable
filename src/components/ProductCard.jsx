import { Link } from "react-router-dom";
import styled from "styled-components";

const Card = styled.div`
  height: 100%;
  padding: 16px;
  border: 2px solid black;
`;

const ProductLink = styled(Link)`
  text-decoration: none;
  color: inherit;
  height: 100%;
  display: flex;
  flex-direction: column;
`;

const ImageContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 16px;
  flex: 1;

  & img {
    width: 100%;
    max-width: 160px;
  }
`;

const ProductTitle = styled.p`
  font-weight: 500;

  display: -webkit-box;
  line-clamp: 3;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;

  line-height: 1.5em;
  height: 4.5em; /* 3 lines */
`;

const ProductCard = ({ product }) => {
  return (
    <Card>
      <ProductLink to={`/product/${product.id}`} title={product.title}>
        <ImageContainer>
          <img src={product.image} alt={product.title} />
        </ImageContainer>
        <ProductTitle>{product.title}</ProductTitle>
        <p>${product.price}</p>
      </ProductLink>
    </Card>
  );
};

export default ProductCard;
