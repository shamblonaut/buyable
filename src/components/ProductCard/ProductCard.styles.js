import { Link } from "react-router-dom";
import styled from "styled-components";

export const Card = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  overflow: hidden;
  height: 480px;

  color: var(--color-light);
  background-color: var(--color-accent);
  border: 3px solid var(--color-dark);
  border-radius: 8px;
  padding-bottom: 8px;
`;

export const ProductLink = styled(Link)`
  text-decoration: none;
  color: inherit;

  width: 100%;
  display: flex;
  flex-direction: column;
  flex: 1;
`;

export const ImageContainer = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 16px;
  flex: 1;

  background-color: white;
  border-bottom: 3px solid var(--color-dark);

  & img {
    max-width: 160px;
    max-height: 200px;
  }
`;

export const ProductDetails = styled.div`
  padding: 16px;
  padding-bottom: 0;
`;

export const ProductTitle = styled.p`
  font-size: 1.25rem;
  font-weight: 700;

  display: -webkit-box;
  line-clamp: 2;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
  text-overflow: ellipsis;
  line-height: 1.5em;
  height: 3em; /* 2 lines */
`;

export const ProductPrice = styled.p`
  font-family: var(--font-highlight);
  font-size: 1.25rem;
  font-weight: 500;
  margin: 8px 0;
`;
