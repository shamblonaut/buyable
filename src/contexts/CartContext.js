import { createContext } from "react";

const CartContext = createContext({
  cart: {},
  dispatchCartAction: () => {},
});

export default CartContext;
