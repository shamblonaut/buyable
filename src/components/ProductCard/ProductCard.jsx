import PropTypes from "prop-types";

import { ProductType } from "@/utils/types";

import {
  Card,
  ProductLink,
  ImageContainer,
  ProductDetails,
  ProductTitle,
  ProductPrice,
} from "./ProductCard.styles";

import { CartActions } from "@/components";

const ProductCard = ({ product, cart, setCart }) => {
  return (
    <Card>
      <ProductLink to={`/product/${product.id}`} title={product.title}>
        <ImageContainer>
          <img src={product.image} alt={product.title} />
        </ImageContainer>
        <ProductDetails>
          <ProductTitle>{product.title}</ProductTitle>
          <ProductPrice>${product.price}</ProductPrice>
        </ProductDetails>
      </ProductLink>
      <CartActions product={product} cart={cart} setCart={setCart} />
    </Card>
  );
};

ProductCard.propTypes = {
  product: ProductType.isRequired,
  cart: PropTypes.object.isRequired,
  setCart: PropTypes.func.isRequired,
};

export default ProductCard;
