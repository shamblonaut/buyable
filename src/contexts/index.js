import CartContext from "./CartContext";

export { CartContext };
