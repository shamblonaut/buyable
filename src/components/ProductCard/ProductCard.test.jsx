import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";

import { BrowserRouter } from "react-router-dom";

import ProductCard from "./ProductCard";
import { CartContext } from "@/contexts";

const mockProduct = {
  id: 1,
  title: "Mock Product",
  price: 15.99,
  image: "http://example.com/mock-image.jpg",
};
const mockCart = { 1: 16 };
const mockSetCart = vi.fn();

vi.mock("./ProductCard.styles", async (importOriginal) => {
  const actual = await importOriginal();

  const { ProductLink } = actual;
  return {
    ...actual,
    ProductLink: ({ children, ...props }) => (
      <ProductLink data-testid="product-link" {...props}>
        {children}
      </ProductLink>
    ),
  };
});

describe("ProductCard component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <CartContext.Provider value={{ cart: mockCart, setCart: mockSetCart }}>
          <ProductCard product={mockProduct} />
        </CartContext.Provider>
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it("links to product page of given product", () => {
    render(
      <BrowserRouter>
        <CartContext.Provider value={{ cart: mockCart, setCart: mockSetCart }}>
          <ProductCard product={mockProduct} />
        </CartContext.Provider>
      </BrowserRouter>,
    );

    expect(screen.getByTestId("product-link")).toHaveAttribute(
      "href",
      `/product/${mockProduct.id}`,
    );
  });
});
