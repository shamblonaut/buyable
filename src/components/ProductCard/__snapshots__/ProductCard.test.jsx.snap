// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard component > renders correctly 1`] = `
<div>
  <div
    class="sc-ibashp bFfuKz"
  >
    <a
      aria-label="product-link"
      class="sc-blIAwI cqQHhd"
      data-discover="true"
      href="/product/1"
      title="Mock Product"
    >
      <div
        class="sc-itBLYH kHAjvE"
      >
        <img
          alt="Mock Product"
          src="http://example.com/mock-image.jpg"
        />
      </div>
      <div
        class="sc-bEjUoa kylyIU"
      >
        <p
          class="sc-boKDdR bKlUwP"
        >
          Mock Product
        </p>
        <p
          class="sc-fOOuSg eGlDie"
        >
          $
          15.99
        </p>
      </div>
    </a>
    <div
      class="sc-kjwdDK dQWvrC"
    >
      <div
        aria-label="cart-quantity-selector"
        class="sc-jytpVa dZPxi"
      >
        <button
          aria-label="quantity-decrement-button"
          class="sc-eknHtZ sc-pYNGo cymyjv hGBICG"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-minus"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
          </svg>
        </button>
        <input
          aria-label="custom-quantity-input"
          class="sc-jYczwO lfFPYm"
          type="text"
          value="16"
        />
        <button
          aria-label="quantity-increment-button"
          class="sc-eknHtZ sc-cCVJLD cymyjv iGXDQy"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-plus"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12h14"
            />
            <path
              d="M12 5v14"
            />
          </svg>
        </button>
      </div>
      <button
        aria-label="cart-remove-button"
        class="sc-eVqvcJ sc-cdmAjP dyuNnT fawTpi"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-trash2 lucide-trash-2"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 6h18"
          />
          <path
            d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
          />
          <path
            d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
